File: apigateway/websocket/websocket_send_msg.js
const AWS = require('aws-sdk');

const ddb = new AWS.DynamoDB.DocumentClient({ apiVersion: '2012-08-10' });

const { TableName } = process.env;

exports.lambda_handler = async (event, context) => {
  // console.log(`Environment Variables:\n` + JSON.stringify(process.env, null, 2));
  // console.log(`WebSocket sendmsg: DynamoDB Table: ${TableName}`);
  console.log(`WebSocket sendmsg: event argument:\n` + JSON.stringify(event,null,2));

  // Retrieve all current connections
  let connectionData;
  try {
    connectionData = await ddb.scan({ TableName: TableName,
                                      ProjectionExpression: 'connectionId' }).promise();
  } catch (e) {
    console.log(e.stack);
    return { statusCode: 500, body: e.stack };
  }

  // Create management object for posting the message to each connection
  const apigwManagementApi = new AWS.ApiGatewayManagementApi({
    apiVersion: '2018-11-29',
    endpoint: event.requestContext.domainName + '/' + event.requestContext.stage
  });

  // Retrieve the message
  const message = JSON.parse(event.body).msg;
  console.log(`WebSocket sendmsg: Received message "${message}"`);

  // Send the message to each connection
  const postCalls = connectionData.Items.map(async ({ connectionId }) => {
    try {
      await apigwManagementApi.postToConnection({ ConnectionId: connectionId, Data: message }).promise();
    } catch (e) {
      if (e.statusCode === 410) {
        console.log(`Found stale connection, deleting ${connectionId}`);
        await ddb.delete({ TableName: TableName, Key: { connectionId } }).promise();
      } else {
        throw e;
      }
    }
  });

  try {
    await Promise.all(postCalls);
  } catch (e) {
    return { statusCode: 500, body: e.stack };
  }

  return { statusCode: 200, body: 'Data sent.' };
};

File: apigateway/websocket/websocket_send_msg.py
import json
import logging
import os
import boto3
from botocore.exceptions import ClientError

# Set up logging
logging.basicConfig(format='%(levelname)s: %(asctime)s: %(message)s')
logger = logging.getLogger()
logger.setLevel(logging.INFO)


def lambda_handler(event, context):
    """Example WebSocket customRoute Lambda function

    The function is called when the route selection expression
    $request.body.action matches "sendmsg", as in:

        {"action": "sendmsg", "msg": "Hello"}

    :param event: Dict (usually) of parameters passed to the function
    :param context: LambdaContext object of runtime data
    :return: Dict of key:value pairs
    """

    # Log the values received in the event and context arguments
    logger.info('message event: ' + json.dumps(event, indent=2))
    connectionId = event['requestContext']['connectionId']
    logger.info(f'message event["requestContext"]["connectionId"]: {connectionId}')

    # Retrieve the name of the DynamoDB table to store connection IDs
    table_name = os.environ['TableName']

    # Retrieve the user name for the connection that's sending the message
    user_name = 'Anon'
    dynamodb_client = boto3.client('dynamodb')
    key = {'connectionId': {'S': connectionId}}
    try:
        response = dynamodb_client.get_item(TableName=table_name,
                                            Key=key,
                                            ProjectionExpression='userName')
    except ClientError as e:
        logger.error(e)
        # raise ValueError(e)
    else:
        user_name = response['Item']['userName']['S']

    # Retrieve all connection IDs from the table
    try:
        response = dynamodb_client.scan(TableName=table_name,
                                        ProjectionExpression='connectionId')
    except ClientError as e:
        logger.error(e)
        raise ValueError(e)

    # Construct the message text as bytes
    body = json.loads(event['body'])
    message = f'{user_name}: {body["msg"]}'.encode('utf-8')
    logger.info(f'Message: "{message}"')

    # Send the message to each connection
    api_client = boto3.client('apigatewaymanagementapi')
    for item in response['Items']:
        connectionId = item['connectionId']['S']
        try:
            api_client.post_to_connection(Data=message,
                                          ConnectionId=connectionId)
        except ClientError as e:
            logger.error(e)

    # Construct response
    response = {'statusCode': 200}
    return response
